### PR TITLE
add sys_utimensat

### PR DESCRIPTION
**Pull request by xiao-mang**
## 拉取/合并请求描述：(PR description)

[

#### 为什么提交这份PR (why to submit this PR)

The lack of sys_utimensat implementation results in an error in the 'touch' command

#### 你的解决方案是什么 (what is your solution)

add sys_utimensat

#### 在什么测试环境下测试通过 (what is the test environment)

qemu aarch64
]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [x] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

**Comment by xqyjlj** [comment]
需要对接tv_nsec ，参考标准 https://man7.org/linux/man-pages/man2/utimensat.2.html

```
If the tv_nsec field of one of the timespec structures has the
special value UTIME_NOW, then the corresponding file timestamp is
set to the current time.  If the tv_nsec field of one of the
timespec structures has the special value UTIME_OMIT, then the
corresponding file timestamp is left unchanged.  In both of these
cases, the value of the corresponding tv_sec field is ignored.
```

File: components/dfs/dfs_v2/include/dfs_fs.h [hunk 1]
```diff
@@ -31,7 +31,10 @@ struct dfs_partition
 
 struct dfs_attr
 {
+    unsigned int	ia_valid;
     mode_t  st_mode;
+    struct timespec ia_atime;
+	struct timespec ia_mtime;
```
Comment (geniusgogo) [inline components/dfs/dfs_v2/include/dfs_fs.h:37]: tab健

File: components/lwp/lwp_syscall.c [hunk 2]
```diff
@@ -6661,6 +6700,7 @@ const static struct rt_syscall_def func_table[] =
     SYSCALL_SIGN(sys_memfd_create),                     /* 200 */
     SYSCALL_SIGN(sys_ftruncate),
     SYSCALL_SIGN(sys_setitimer),
+    SYSCALL_SIGN(sys_utimensat),
```
Comment (geniusgogo) [inline components/lwp/lwp_syscall.c:6703]: 这个系统调用号，没办法和工具链对应起来啊，工具链那边少了个sys_setitimer?

File: components/dfs/dfs_v2/src/dfs_posix.c [hunk 3]
```diff
@@ -125,6 +125,129 @@ int openat(int dirfd, const char *path, int flag, ...)
...
+#ifndef ATTR_ATIME_SET
+#define ATTR_ATIME_SET	(1 << 7)
+#endif
+#ifndef ATTR_MTIME_SET
+#define ATTR_MTIME_SET	(1 << 8)
+#endif
+#ifndef AT_SYMLINK_NOFOLLOW
+#define AT_SYMLINK_NOFOLLOW 0x100
+#endif
+#ifndef GET_ERRNO
```
Comment (geniusgogo) [inline components/dfs/dfs_v2/src/dfs_posix.c:137]: GET_ERRNO、SYS_WARP都是属于syscall的wrapper层负责的事情，不应该出现在其它地方。

**Review by geniusgogo — CHANGES_REQUESTED** [review]

**Review by geniusgogo — APPROVED** [review]

File: components/dfs/dfs_v2/src/dfs_posix.c [hunk 4]
```diff
@@ -125,6 +125,123 @@ int openat(int dirfd, const char *path, int flag, ...)
...
+            if (!fullpath)
+            {
+                rt_set_errno(-ENOMEM);
+                return -1;
+            }
+        }
+    }
+
+    //update time
+    attr.ia_valid = ATTR_ATIME_SET | ATTR_MTIME_SET;
```
Comment (geniusgogo) [inline components/dfs/dfs_v2/src/dfs_posix.c:190]: 这种flag会在什么地方解析判断呢？

Comment (xiao-mang) [reply components/dfs/dfs_v2/src/dfs_posix.c:190]: ext4/ports/rtthread/dfs_ext.c的dfs_ext_setattr函数

**Review by geniusgogo — APPROVED** [review]

**Review by BernardXiong — APPROVED** [review]

**Pull request merged**